<a href="https://colab.research.google.com/github/Manisha021095/Manisha_INFO5731_Fall2022/blob/main/In_class_exercise_02_09222022.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## The second In-class-exercise (09/22/2022, 40 points in total)

The purpose of this exercise is to understand users' information needs, then collect data from different sources for analysis.

Question 1 (10 points): Describe an interesting research question (or practical question) you have in mind, what kind of data should be collected to answer the question(s)? How many data needed for the analysis? The detail steps for collecting and save the data. 

In [ ]:
# Your answer here (no code for this question, write down your answer as detail as possible for the above questions):


'''
Research question - What is the review of Nokia Andriod smart TV. Whether it has positive reviews or negative reviews.
Data required for this research is reviews of the customer who already bought the TV. Ecommerce website contains the reviews 
of the product. I have choosen the flipkart to collect the reviews of the Nokia Andriod smart TV. We can analyze the reviews 
by considering the words in them that is whether the user customer has written in a positive way or negative way. We need 
minimum 1000 reviews to get nearly accurate results.
`
Steps for Collecting and Saving Data:
I have used the BeautifulSoup library to extract the information from the website.
I have extracted the reviews by using the classname and then appended to the empty list.
To extract 1000 reviews I have itearted 100 times as each page contains 10 reviews and I have generated the url dynamically while 
iterating.
Then created a dataframe form the list and then converted dataframe to csv




'''



'\nResearch question - What is the review of Nokia Andriod smart TV. Whether it has positive reviews or negative reviews.\nData required for this research is reviews of the customer who already bought the TV. Ecommerce website contains the reviews \nof the product. I have choosen the flipkart to collect the reviews of the Nokia Andriod smart TV. We can analyze the reviews \nby considering the words in them that is whether the user customer has written in a positive way or negative way. We need \nminimum 500 reviews to get nearly accurate results.\n\nSteps for Collecting and Saving Data:\nI have used the BeautifulSoup library to extract the information from the website.\nI have extracted the reviews by using the classname and then appended to the empty list.\nTo extract 500 reviews I have itearted 50 times as each page contains 10 reviews and I have generated the url dynamically while \niterating.\nThen created a dataframe form the list and then converted dataframe to csv\n\n\n\n\n'

Question 2 (10 points): Write python code to collect 1000 data samples you discussed above.

In [ ]:
# You code here (Please add comments in the code):

import requests
from bs4 import BeautifulSoup
import pandas as pd

main_text = [] # List to store Review headings
sub_text =[] #List to store reviews

for number in range(1,101):
    link = "https://www.flipkart.com/nokia-139cm-55-inch-ultra-hd-4k-led-smart-android-tv-sound-jbl/product-reviews/itmffvfvyztsmfmq?pid=TVSFFVFVJEGZ3R5H&lid=LSTTVSFFVFVJEGZ3R5HW9DJ5W&marketplace=FLIPKART&page=" + str(number) # Generating link dynamically
    page = requests.get(link) # Accessing the webpage
    soup = BeautifulSoup(page.text, 'html.parser')
    main_reviews = soup.find_all(class_='_2-N8zT') # Getting the Review Heading by using the class name
    text_reviews = soup.find_all(class_='t-ZTKy') # Getting the full reviews by using the class name
    for ele, sub_ele in zip(main_reviews, text_reviews) : # Iterating through the list
        main_text.append(ele.text) #Appending to empty list
        sub_text.append(sub_ele.text)
      

In [ ]:
df = pd.DataFrame(list(zip(main_text, sub_text)), columns =['Glimpse of Review', 'Full Review'])  # Creating Dataframe
print("Length of data frame is {0}".format(len(df)))
df

Length of data frame is 1000


Glimpse of Review  \
0    Review from Technology Gyan: Almost everything...   
1                                            Brilliant   
2                                  Best in the market!   
3                                          Pretty good   
4                                    Terrific purchase   
..                                                 ...   
995                                   Perfect product!   
996                               Classy product ,Good   
997                                          Wonderful   
998                                           Not good   
999                                             better   

                                           Full Review  
0    This 55" 4K Nokia TV at this price point comes...  
1    I must say it is best decision to by Nokia TV....  
2    Flipcart delivered the Product in less than 24...  
3    Pros1) Picture Quality is good, micro dimming ...  
4    Value for money ....1) This tv has Everything ...  
..                                                 ...  
995  I have reviewed ofter using 30 days. Tv is rea...  
996  very good TV, Everything fine. Go for it.READ ...  
997  The product does well for a budget TV.Good pic...  
998                                 No thanksREAD MORE  
999                                      goodREAD MORE  

[1000 rows x 2 columns]

In [ ]:
df.to_csv('Question2_data.csv', index=False) #Converting to csv file from dataframe

Question 3 (10 points): Write python code to collect 1000 articles from Google Scholar (https://scholar.google.com/), Microsoft Academic (https://academic.microsoft.com/home), or CiteSeerX (https://citeseerx.ist.psu.edu/index), or Semantic Scholar (https://www.semanticscholar.org/), or ACM Digital Libraries (https://dl.acm.org/) with the keyword "information retrieval". The articles should be published in the last 10 years (2012-2022).

The following information of the article needs to be collected:

(1) Title

(2) Venue/journal/conference being published

(3) Year

(4) Authors

(5) Abstract

In [ ]:
# You code here (Please add comments in the code):

key_word = "information retrieval"


key_word_format = "+".join(key_word.split(" "))


check_link ="https://dl.acm.org/action/doSearch?AllField=" + key_word_format + "&pageSize=50&startPage=" 


In [ ]:
Req_df = pd.DataFrame(columns = ["Title", "Venue/Journal/Conf", "Year", "Authors", "Abstract"])

for i in range(1,200):
    try:
        search_link = check_link + str(i)
        page = requests.get(search_link)
        soup = BeautifulSoup(page.text, 'html.parser')
        titles = soup.find_all(class_ = "hlFld-Title")
        for j in titles:
            if len(Req_df) < 1001:       
                sub_link = "https://dl.acm.org" + j.find("a").get("href")
                sub_page = requests.get(sub_link)
                sub_soup = BeautifulSoup(sub_page.text, 'html.parser')
                year_j = sub_soup.find_all(class_ = "CitationCoverDate")[0].text.split(" ")[-1]
                if 2011 < int(year_j) < 2023:
                    Req_df = Req_df.append({
                        "Title"              : sub_soup.find_all(class_ = "citation__title")[0].text,
                        "Venue/Journal/Conf" : sub_soup.find_all(class_ = "issue-item__detail")[0].find("a").get("title"),
                        "Year"               : year_j,
                        "Authors"            : sub_soup.find_all(class_ = "author-data")[0].text,
                        "Abstract"           : sub_soup.find_all(class_ = "hlFld-Abstract")[0].text[8:].replace("\n", "")
                    }, ignore_index = True)
                #sub_page.close()
            else:
                break
    except:
        print("page: ", i)
    #page.close()

page:  37
page:  43


In [ ]:
Req_df

Title  \
0    Interactive Information Retrieval with Bandit ...   
1    Proactive Information Retrieval via Screen Sur...   
2               Private Stateful Information Retrieval   
3              Crowdsourcing for information retrieval   
4    Bandit Algorithms in Interactive Information R...   
..                                                 ...   
523  Report on the SIGIR 2013 workshop on modeling ...   
524  Effective Information Retrieval Approach based...   
525  A user defined taxonomy of factors that divide...   
526  Report on the fifth workshop on exploiting sem...   
527  Towards an Optimal Dialog Strategy for Informa...   

                                    Venue/Journal/Conf  Year  \
0    SIGIR '21: Proceedings of the 44th Internation...  2021   
1    SIGIR '17: Proceedings of the 40th Internation...  2017   
2    CCS '18: Proceedings of the 2018 ACM SIGSAC Co...  2018   
3                                      ACM SIGIR Forum  2012   
4    ICTIR '17: Proceedings of the ACM SIGIR Intern...  2017   
..                                                 ...   ...   
523                                    ACM SIGIR Forum  2013   
524  ICARCSET '15: Proceedings of the 2015 Internat...  2015   
525  IIiX '14: Proceedings of the 5th Information I...  2014   
526                                    ACM SIGIR Forum  2012   
527  IUI '18: 23rd International Conference on Inte...  2018   

                 Authors                                           Abstract  
0          Huazheng Wang  Information retrieval (IR) in nature is a proc...  
1             Tung Vuong  We demonstrate proactive information retrieval...  
2           Sarvar Patel  Private information retrieval (PIR) is a funda...  
3          Matthew Lease  The 2nd SIGIR Workshop on Crowdsourcing for In...  
4        Dorota Glowacka  The multi-armed bandit problem models an agent...  
..                   ...                                                ...  
523  Charles L.A. Clarke  The SIGIR 2013 Workshop on Modeling User Behav...  
524       I. Govindharaj  The Information technology is developing rapid...  
525            Chaoyu Ye  Although research is increasingly interested i...  
526           Jaap Kamps  There is an increasing amount of structure on ...  
527        Yunfeng Zhang  The emerging paradigm of dialogue interfaces f...  

[528 rows x 5 columns]

I have selected "ACM Digital Libraries" website to retrieve articles. There are only 528 articles with "information retrieval" keyword from last 10 years.

Question 4 (10 points): Write python code to collect 1000 posts from Twitter, or Facebook, or Instagram. You can either use hashtags, keywords, user_name, user_id, or other information to collect the data. 

The following information needs to be collected:

(1) User_name

(2) Posted time

(3) Text 

In [ ]:
# You code here (Please add comments in the code):

!pip install snscrape
import snscrape.modules.twitter as sntwitter

In [ ]:
attributes_container = []

In [ ]:
for i,tweet in enumerate(sntwitter.TwitterSearchScraper('data since:2021-01-01 until:2022-09-20').get_items()):
    if i>1000:
        break
    attributes_container.append([tweet.user.username, tweet.date, tweet.content])

In [ ]:
tweets_df = pd.DataFrame(attributes_container, columns=["User", "Date Created", "Tweet"])
tweets_df

User              Date Created  \
0      tigerladytexas 2022-09-19 23:59:59+00:00   
1          davesnyder 2022-09-19 23:59:59+00:00   
2        ProdukPantau 2022-09-19 23:59:59+00:00   
3           MLB_DATA_ 2022-09-19 23:59:59+00:00   
4     cereja_stranger 2022-09-19 23:59:59+00:00   
...               ...                       ...   
996       data_diario 2022-09-19 23:39:22+00:00   
997     curtisbuzzard 2022-09-19 23:39:21+00:00   
998      bookstand171 2022-09-19 23:39:21+00:00   
999      info_daratan 2022-09-19 23:39:20+00:00   
1000    homo10sapiens 2022-09-19 23:39:20+00:00   

                                                  Tweet  
0     @RoxineKing @AZ_Brittney Where did they get th...  
1     @bernardjhuang Be able to build briefs as part...  
2     Coba cek ini, deh: "UNEED Nylon Kabel Data Typ...  
3     Julio Teheran(ATL) Age.24\n33登板11勝8敗\n防御率4.04\...  
4     Todo dia uma foto aleatória até falarem a data...  
...                                                 ...  
996   💥 @CFKArgentina, sobre la Causa Vialidad: "Vam...  
997   Great session on data literacy at the Maneuver...  
998   fetterman of pennsylvania and biden of delawar...  
999   Idrus memuji kinerja KPU khususnya terkait kew...  
1000  A pesquisa data Povo é real! E mostra outra re...  

[1001 rows x 3 columns]